File: Algorithms/0543_Diameter_of_Binary_Tree/Python/Diameter_of_Binary_Tree_Solution_1.py

# Space: O(1)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root):
        self.res = 0
        self.dfs(root)
        return self.res

    def dfs(self, root):
        if root is None: return 0
        left, right = 0, 0
        if root.left:
            left = self.dfs(root.left) + 1
        if root.right:
            right = self.dfs(root.right) + 1
        self.res = max(self.res, left + right)
        return max(left, right)

File: Algorithms/0543_Diameter_of_Binary_Tree/Java/DiameterofBinaryTreeSolution1.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int res = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return res;
    }

    private int dfs(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return 0;
        int left = 0, right = 0;
        if (root.left != null) left = dfs(root.left)+1;
        if (root.right != null) right = dfs(root.right)+1;
        res = Math.max(res, left + right);
        return Math.max(left, right);

    }
}




